File: app/src/main/java/com/spiderbiggen/randomchampionselector/activities/ListChampionsActivity.java
package com.spiderbiggen.randomchampionselector.activities;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;

import java.util.ArrayList;

import com.spiderbiggen.randomchampionselector.R;
import com.spiderbiggen.randomchampionselector.model.Champion;
import com.spiderbiggen.randomchampionselector.storage.database.DatabaseManager;
import com.spiderbiggen.randomchampionselector.view.adapters.ChampionAdapter;

import io.reactivex.disposables.Disposable;

public class ListChampionsActivity extends ButtonActivity {

    private final ChampionAdapter adapter;
    private RecyclerView recyclerView;
    private Disposable listFlowable;

    public ListChampionsActivity() {
        adapter = new ChampionAdapter(new ArrayList<>(), this::onClick);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        setContentView(R.layout.activity_list_champions);
        super.onCreate(savedInstanceState);
        recyclerView = findViewById(R.id.champion_list);
        recyclerView.setAdapter(adapter);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setTitle(getTitle());
        }
    }

    @Override
    protected void onResume() {
        openChampionList();
        super.onResume();
    }

    @Override
    protected void onPause() {
        dispose();
        super.onPause();
    }

    public void openChampionList() {
        dispose();
        listFlowable = DatabaseManager.getInstance().findChampionList(adapter::setChampions, null);
    }

    private void dispose() {
        if (listFlowable != null && !listFlowable.isDisposed()) {
            listFlowable.dispose();
        }
    }

    private void onClick(View v) {
        int position = recyclerView.getChildAdapterPosition(v);
        Champion champion = adapter.getChampion(position);
        ImageView img = v.findViewById(R.id.champion_splash);

        Intent intent = getChampionIntent();
        intent.putExtra(ChampionActivity.CHAMPION_KEY, champion.getKey());
        ActivityOptionsCompat options = ActivityOptionsCompat
            .makeSceneTransitionAnimation(this, img, getString(R.string.champion_splash_transition_key));
        intent.putExtra(ChampionActivity.UP_ON_BACK_KEY, false);
        startActivity(intent, options.toBundle());
    }
}

File: app/src/main/java/com/spiderbiggen/randomchampionselector/activities/ListChampionsActivity.kt
package com.spiderbiggen.randomchampionselector.activities

import android.app.ActivityOptions
import android.content.Intent
import android.os.Bundle
import android.util.Pair
import android.view.Menu
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.activity.viewModels
import com.spiderbiggen.randomchampionselector.DataApplication
import com.spiderbiggen.randomchampionselector.R
import com.spiderbiggen.randomchampionselector.adapters.ChampionAdapter
import com.spiderbiggen.randomchampionselector.databinding.ActivityListChampionsBinding
import com.spiderbiggen.randomchampionselector.viewmodels.ChampionListViewModel
import com.spiderbiggen.randomchampionselector.viewmodels.ChampionListViewModelFactory
import kotlinx.coroutines.*

class ListChampionsActivity : AbstractActivity() {
    private lateinit var binding: ActivityListChampionsBinding

    private val adapter = ChampionAdapter(View.OnClickListener(this::onClick))

    private val viewModel: ChampionListViewModel by viewModels {
        ChampionListViewModelFactory((application as DataApplication).championRepository)
    }

    @ExperimentalCoroutinesApi
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityListChampionsBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setSupportActionBar(binding.toolbar)
        supportActionBar?.title = title
        binding.championList.adapter = adapter
        binding.activity = this
        viewModel.champions.observe(this, {
            viewModel.selectRandomChampion()
            adapter.setChampions(it)
        })
        viewModel.bitmap.observe(this, { binding.splash.setImageBitmap(it) })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu, menu)
        return true
    }

    fun openChampion(view: View) {
        openChampionActivity(view)
    }

    private fun onClick(v: View) {
        val position = binding.championList.getChildAdapterPosition(v)
        val img = v.findViewById<ImageView>(R.id.champion_splash)
        val name = v.findViewById<TextView>(R.id.champion_name)
        val title = v.findViewById<TextView>(R.id.champion_title)
        val champion = adapter.getChampion(position)
        val intent = Intent(this, ChampionActivity::class.java)
        intent.putExtra(ChampionActivity.CHAMPION_KEY, champion?.key)
        val options = ActivityOptions.makeSceneTransitionAnimation(
            this,
            Pair(img, getString(R.string.champion_splash_transition_key)),
            Pair(name, getString(R.string.champion_name_transition_key)),
            Pair(title, getString(R.string.champion_title_transition_key))
        )
        intent.putExtra(ChampionActivity.UP_ON_BACK_KEY, false)
        startActivityWithFade(intent, options.toBundle())
    }
}
